src/hotspot/share/runtime/synchronizer.cpp

```diff
@@ -2336,185 +1389,55 @@ class HandshakeForDeflation : public HandshakeClosure {
-      // and then unmarked while prepend_to_common() is sorting it
-      // all out.
-#ifdef ASSERT
-      ObjectMonitor* l_next_om = unmarked_next(free_tail_p);
-      assert(l_next_om == NULL, "must be NULL: _next_om=" INTPTR_FORMAT, p2i(l_next_om));
-#endif
-
-      prepend_list_to_global_wait_list(free_head_p, free_tail_p, local_deflated_count);
-
-      OM_PERFDATA_OP(Deflations, inc(local_deflated_count));
-    }
-
-    if (saved_mid_in_use_p != NULL) {
-      // deflate_monitor_list_using_JT() detected a safepoint starting.
-      timer.stop();
-      {
-        if (is_global) {
-          log_debug(monitorinflation)("pausing deflation of global idle monitors for a safepoint.");
-        } else {
-          log_debug(monitorinflation)("jt=" INTPTR_FORMAT ": pausing deflation of per-thread idle monitors for a safepoint.", p2i(target));
-        }
-        assert(SafepointMechanism::should_process(self), "sanity check");
-        ThreadBlockInVM blocker(self);
-      }
-      // Prepare for another loop after the safepoint.
-      free_head_p = NULL;
-      free_tail_p = NULL;
-      if (log_is_enabled(Info, monitorinflation)) {
-        timer.start();
+  // Deflate any global idle monitors.
+  size_t deflated = deflate_monitor_list();
+  if (deflated > 0) {
+    // There are monitors that have been deflated. Unlink them from the in use list.
+    ResourceMark rm;
+    GrowableArray<ObjectMonitor*> delete_list((int)deflated);
+    _in_use_list.unlink_deflated(&delete_list);
+    HandshakeForDeflation hfd_hc;
+    Handshake::execute(&hfd_hc);
+    // After the handshake, safely free the monitors that were deflated in this cycle.
+    for (ObjectMonitor* monitor: delete_list) {
+      delete monitor;
+      if (SafepointMechanism::should_process(self)) {
+        // Honor block request.
+        ThreadBlockInVM tbivm(self);
       }
     }
-  } while (saved_mid_in_use_p != NULL);
+  }
   timer.stop();
 
+  OM_PERFDATA_OP(MonExtant, set_value(_in_use_list.count()));
+  OM_PERFDATA_OP(Deflations, inc(deflated));
+
   LogStreamHandle(Debug, monitorinflation) lsh_debug;
   LogStreamHandle(Info, monitorinflation) lsh_info;
   LogStream* ls = NULL;
   if (log_is_enabled(Debug, monitorinflation)) {
     ls = &lsh_debug;
-  } else if (deflated_count != 0 && log_is_enabled(Info, monitorinflation)) {
+  } else if (deflated != 0 && log_is_enabled(Info, monitorinflation)) {
     ls = &lsh_info;
   }
   if (ls != NULL) {
-    if (is_global) {
-      ls->print_cr("async-deflating global idle monitors, %3.7f secs, %d monitors", timer.seconds(), deflated_count);
-    } else {
-      ls->print_cr("jt=" INTPTR_FORMAT ": async-deflating per-thread idle monitors, %3.7f secs, %d monitors", p2i(target), timer.seconds(), deflated_count);
-    }
+    ls->print_cr("deflating idle monitors, %3.7f secs, " SIZE_FORMAT " monitors", timer.seconds(), deflated);
   }
+  log_info(monitorinflation)("global_population=" SIZE_FORMAT, _in_use_list.count());
```

Comment:
This should change from `global_population` to `population` and the value
should be the population value computed by:
    `AvgMonitorsPerThreadEstimate * n_JavaThreads`
We also need to add back the `in_use_count`. The old log message used to be:
    `log_info(monitorinflation)("async global_population=%d, global_in_use_count=%d, "`
Also need to think about what should happen if the in_use_count exceeds the
population count? Should the population count float up? Should it be renamed
from population count to population ceiling?